proposals/binary-compat-only.md

```diff
@@ -0,0 +1,123 @@
+> - **If `M` is marked with `BinaryCompatOnlyAttribute`, the accessibility domain of `M` is completely inaccessible to the program text of `P` and any program that references `P`.**
+> - If the declared accessibility of `M` is `public`, the accessibility domain of `M` is the accessibility domain of `T`.
+> - If the declared accessibility of `M` is `protected internal`, let `D` be the union of the program text of `P` and the program text of any type derived from `T`, which is declared outside `P`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `private protected`, let `D` be the intersection of the program text of `P` and the program text of `T` and any type derived from `T`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `protected`, let `D` be the union of the program text of `T`and the program text of any type derived from `T`. The accessibility domain of `M` is the intersection of the accessibility domain of `T` with `D`.
+> - If the declared accessibility of `M` is `internal`, the accessibility domain of `M` is the intersection of the accessibility domain of `T` with the program text of `P`.
+> - If the declared accessibility of `M` is `private`, the accessibility domain of `M` is the program text of `T`.
+
+The goal of these additions is to make it so that members marked with `BinaryCompatOnlyAttribute` are completely inaccessible to any location, they will
+not participate in member lookup, and cannot affect the rest of the program. Consequentely, this means they cannot implement interface members, they cannot
+call each other, and they cannot be overridden (virtual methods), hidden, or implemented (interface members). Whether this is too strict is the subject of
+several open questions below.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+More complex attribute mechanisms make the language more complex.
+
+## Alternatives
+[alternatives]: #alternatives
+
+No other alternatives have presented themselves
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+### Virtual methods and overriding
+
+What do we do when a virtual method is marked as `BinaryCompatOnly`? Overrides in a derived class may not even be in the current assembly, and it could
+be that the user is looking to introduce a new version of a method that, for example, only differs by return type, something that C# does not normally
+allow overloading on. What happens to any overrides of that previous method on recompile? Are they allowed to override the `BinaryCompatOnly` member if
+they're also marked as `BinaryCompatOnly`?
+
+### Use within the same DLL
+
+This proposal states that `BinaryCompatOnly` members are not visible anywhere, not even in the assembly currently being compiled. Is that too strict, or
+do `BinaryCompatAttribute` members need to possibly chain to one another?
```

Comment:
I think a good solution to this is the following: disallow referencing `BinaryCompatOnly` members (other than types) completely, their implementation can be extracted to a `private` / `internal` member as required which the `BinaryCompatOnly` will stub to. For unit tests, you would either test the stub, or use reflection.

There is a more interesting question with types though, as you still need to be able to use them in signatures probably (even if just other `BinaryCompatOnly` signatures); I think the solution to this could be: only use that type if there's no other type which matches (i.e., force all `BinaryCompatOnly` types to match very last, after every other possibility) (or can be done by specifying full name `global::` if needed - probably the recommended way to do it), and emit a warning on usage of a `BinaryCompatOnly` type (default error severity), which can be suppressed if the usage is intentional. Potential (quite reasonable) adjustment could be that: this only applies for already `BinaryCompatOnly` members/types, it would be disallowed & ignored otherwise.